### PR TITLE
implement Replicated Map with tests

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
#158.

This PR is continuation of #180.

### CODE REVIEW

File: internal/replicated_map.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	partitionCount := client.PartitionService.PartitionCount()
+	targetPartitionId := rand.Int31n(partitionCount)
+	return &ReplicatedMapProxy{proxy: &proxy{client, serviceName, name}, targetPartitionId: targetPartitionId}
+}
+func (rmp *ReplicatedMapProxy) Put(key interface{}, value interface{}) (oldValue interface{}, err error) {
+	keyData, valueData, err := rmp.validateAndSerialize2(key, value)
+	if err != nil {
+		return nil, err
+	}
+	request := ReplicatedMapPutEncodeRequest(rmp.name, keyData, valueData, TTL)
```
Comment (mustafaiman) [inline internal/replicated_map.go]: `TTL` here comes from map.go. This is not good practice. I dont think this constant adds value. Straight `0` would be more readable. If you really want to use a constant, name it better (`TTL_UNLIMITED`) and make it private to this proxy.

File: internal/replicated_map.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	targetPartitionId int32
+}
+
+func newReplicatedMapProxy(client *HazelcastClient, serviceName *string, name *string) *ReplicatedMapProxy {
+	partitionCount := client.PartitionService.PartitionCount()
+	targetPartitionId := rand.Int31n(partitionCount)
+	return &ReplicatedMapProxy{proxy: &proxy{client, serviceName, name}, targetPartitionId: targetPartitionId}
+}
+func (rmp *ReplicatedMapProxy) Put(key interface{}, value interface{}) (oldValue interface{}, err error) {
+	keyData, valueData, err := rmp.validateAndSerialize2(key, value)
```
Comment (mustafaiman) [inline internal/replicated_map.go]: You may reuse `PutWithTtl` here.

File: internal/common/util.go [hunk 3]
```diff
@@ -46,9 +47,18 @@ func GetIpAndPort(addr string) (string, int32) {
 func CheckNotNil(v interface{}) bool {
 	return v != nil
 }
+
+func AssertNotNil(argument interface{}, argName string) error {
```
Comment (mustafaiman) [inline internal/common/util.go]: I dont see this function being used in this PR.

File: internal/replicated_map.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+func (rmp *ReplicatedMapProxy) EntrySet() (resultPairs []IPair, err error) {
+	request := ReplicatedMapEntrySetEncodeRequest(rmp.name)
+	responseMessage, err := rmp.InvokeOnPartition(request, rmp.targetPartitionId)
+	return rmp.DecodeToPairSliceAndError(responseMessage, err, ReplicatedMapEntrySetDecodeResponse)
+}
+
+func (rmp *ReplicatedMapProxy) AddEntryListener(listener interface{}) (registrationID *string, err error) {
+	request := ReplicatedMapAddEntryListenerEncodeRequest(rmp.name, rmp.isSmart())
+	eventHandler := func(clientMessage *ClientMessage) {
```
Comment (mustafaiman) [inline internal/replicated_map.go]: can you extract event handler creation out of this method?

File: internal/replicated_map.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+	})
+}
+
+func (rmp *ReplicatedMapProxy) AddEntryListenerWithPredicate(listener interface{}, predicate IPredicate) (registrationID *string, err error) {
+	predicateData, err := rmp.validateAndSerializePredicate(predicate)
+	if err != nil {
+		return nil, err
+	}
+	request := ReplicatedMapAddEntryListenerWithPredicateEncodeRequest(rmp.name, predicateData, rmp.isSmart())
+	eventHandler := func(clientMessage *ClientMessage) {
```
Comment (mustafaiman) [inline internal/replicated_map.go]: extract event handler creation out of this method

File: tests/proxy/map/map_test.go [hunk 6]
```diff
@@ -42,14 +43,24 @@ func TestMain(m *testing.M) {
...
 	m.Run()
-	mp.Clear()
+	clear()
 	client.Shutdown()
 	remoteController.ShutdownCluster(cluster.ID)
 }
 
+func initialize() {
+	mp, _ = client.GetMap("myMap")
+	rmp, _ = client.GetReplicatedMap("myReplicatedMap")
```
Comment (mustafaiman) [inline tests/proxy/map/map_test.go]: `rmp` is not used anywhere in this test. Should be removed.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+	client, _ = hazelcast.NewHazelcastClient()
+	rmp, _ = client.GetReplicatedMap("myReplicatedMap")
+	m.Run()
+	rmp.Clear()
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+}
+
+func TestReplicatedMapProxy_Name(t *testing.T) {
+	name := "myReplicatedMap"
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: no need for this variable. In fact, eliminating the variable would make the test more readable.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+	AssertEqualf(t, nil, name, rmp.Name(), "Name() failed")
+}
+
+func TestReplicatedMapProxy_ServiceName(t *testing.T) {
+	serviceName := common.SERVICE_NAME_REPLICATED_MAP
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: no need for `serviceName` variable. Use the name directly in the assertion.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+	AssertEqualf(t, nil, serviceName, rmp.ServiceName(), "ServiceName() failed")
+}
+
+func TestReplicatedMapProxy_PartitionKey(t *testing.T) {
+	name := "myReplicatedMap"
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: again no need for the variable.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+	defer rmp.Clear()
+	testMap := make(map[interface{}]interface{})
+	for i := 0; i < 10; i++ {
+		testMap["testingKey"+strconv.Itoa(i)] = "testingValue" + strconv.Itoa(i)
+	}
+	err := rmp.PutAll(testMap)
+	AssertErrorNil(t, err)
+	entryList, err := rmp.EntrySet()
+	AssertErrorNil(t, err)
+	for _, pair := range entryList {
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: if `entryList` did not include all the required elements, this test would still pass.
It will be better if you write a utility method that will check equality of two slices and use that. Another utility method may also be necessary for getting the slice from map.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+func TestReplicatedMapProxy_Values(t *testing.T) {
+	defer rmp.Clear()
+	var expecteds = make([]string, 10)
+	var ret = make([]string, 10)
+	for i := 0; i < 10; i++ {
+		rmp.Put(strconv.Itoa(i), strconv.Itoa(i))
+		expecteds[i] = strconv.Itoa(i)
+	}
+	values, err := rmp.Values()
+	for j := 0; j < 10; j++ {
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: what if `values` was broken so that it returned more elements than 10. This test would still pass. It is better if you write a utility method for this kind of comparisons and use it in these kind of tests. Something like `AssertIterableEqual`.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+func TestReplicatedMapProxy_AddEntryListenerWithPredicate(t *testing.T) {
+	var wg *sync.WaitGroup = new(sync.WaitGroup)
+	entryListener := &entryListener{wg: wg}
+	registrationId, err := rmp.AddEntryListenerWithPredicate(entryListener, core.Equal("this", "value"))
+	AssertEqual(t, err, nil, nil)
+	wg.Add(1)
+	rmp.Put("key123", "value")
+	timeout := WaitTimeout(wg, Timeout)
+	AssertEqualf(t, nil, false, timeout, "AddEntryListenerWithPredicate failed")
+	rmp.RemoveEntryListener(registrationId)
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: you dont assert the values that entry listener got here.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+func TestReplicatedMapProxy_AddEntryListenerToKey(t *testing.T) {
+	var wg *sync.WaitGroup = new(sync.WaitGroup)
+	entryListener := &entryListener{wg: wg}
+	registrationId, err := rmp.AddEntryListenerToKey(entryListener, "key1")
+	AssertEqual(t, err, nil, nil)
+	wg.Add(1)
+	rmp.Put("key1", "value1")
+	timeout := WaitTimeout(wg, Timeout)
+	AssertEqualf(t, nil, false, timeout, "AddEntryListenerToKey failed")
```
Comment (mustafaiman) [inline tests/proxy/replicated_map/replicated_map_test.go]: the same as `AddEntryListenerWithPredicate` above.

File: tests/proxy/replicated_map/replicated_map_test.go (hunk 7 above)
Comment (SaitTalhaNisanci) [reply tests/proxy/replicated_map/replicated_map_test.go]: @mustafaiman  I think these tests are not necessary actually.(Name,ServiceName,PartitionKey). They are already tested here https://github.com/hazelcast/hazelcast-go-client/blob/master/tests/proxy/map/map_test.go#L53. Do you think they are necessary ?

File: tests/proxy/map/map_test.go [hunk 14]
```diff
@@ -32,6 +32,7 @@ import (
 
 var mp IMap
 var mp2 IMap
+var rmp ReplicatedMap
```
Comment (mustafaiman) [inline tests/proxy/map/map_test.go]: not used

File: tests/proxy/map/map_test.go [hunk 15]
```diff
@@ -42,14 +43,23 @@ func TestMain(m *testing.M) {
...
 	remoteController.ShutdownCluster(cluster.ID)
 }
 
+func initialize() {
+	mp, _ = client.GetMap("myMap")
+}
+
+func clear() {
+	mp.Clear()
+	rmp.Clear()
```
Comment (mustafaiman) [inline tests/proxy/map/map_test.go]: not used

File: tests/proxy/replicated_map/replicated_map_test.go (hunk 7 above)
Comment (mustafaiman) [reply tests/proxy/replicated_map/replicated_map_test.go]: Agree, not really neccessary.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+	testKey := "testingKey"
+	testValue := "testingValue"
+	rmp.Put(testKey, testValue)
+	rmp.Destroy()
+	rmp, _ := client.GetReplicatedMap("myReplicatedMap")
+	res, err := rmp.Get(testKey)
+	AssertNilf(t, err, res, "replicatedMap Destroy() failed")
+}
+
+func TestReplicatedMapProxy_Put(t *testing.T) {
```
Comment (asimarslan) [inline tests/proxy/replicated_map/replicated_map_test.go]: There are missing tests like TestX_withNil tests

Comment (SaitTalhaNisanci) [reply tests/proxy/replicated_map/replicated_map_test.go]: @asimarslan  I dont think we need those tests here, they go to the same method which is already tested.

Comment (asimarslan) [reply tests/proxy/replicated_map/replicated_map_test.go]: Those validation methods are tested but your proxy method may not use it correctly so We need to test if a proxy method reject a nil value or not. So we do need these tests.

File: tests/proxy/replicated_map/replicated_map_test.go [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+func TestReplicatedMapProxy_Get(t *testing.T) {
+	defer rmp.Clear()
+	testKey := "testingKey"
+	testValue := "testingValue"
+	_, err := rmp.Put(testKey, testValue)
+	res, err := rmp.Get(testKey)
+	AssertEqualf(t, err, res, testValue, "replicatedMap Get() failed")
+}
+
+func TestReplicatedMapProxy_ContainsKey(t *testing.T) {
```
Comment (asimarslan) [inline tests/proxy/replicated_map/replicated_map_test.go]: We also need test with different values. The methods that take values should be tested for both exist and not exists  cases.
Test Contain for an empty map and/or Remove a non exists key.
See queue_test.go
